readme.md

```diff
@@ -26,65 +21,589 @@ or
+  },
+];
+
+// the root of the schema is an array, so is the root of the error
+const expectedError = JSON.stringify([
+  {
+    "index":1,
+    "error": {
+      "name":"Expected string but got undefined.",
+      "address": {
+        "street2":"Expected string but got undefined."
+      }
+    }
+  }
+]);
+expect(() => userSchema(userSchema, value2).toThrow(expectedError);
+```
 ___
 
 ## Core concepts
 #### Decoders
+Decoder is a function that parses unkown value, validates it and returns a known value. A decoder has following signature.
 
+```typescript
+(val: unknown, context?: Context | undefined) => T
+```
+#### Context
+```typescript
+type Context = {
+    key?: string | undefined;
+    index?: number | undefined;
+    schema?: any;
+}
+```
 
 #### Compose complex decoder
 
 #### Validators
+Validator is a function that validates a known value as its name suggests . It returns an error if a value is invalid else returns nothing. A decoder uses 1 or many validators to validate a known value.
 
+```typescript
+(val: T, context?: Context | undefined) => string | Record<string, string> | undefined
+```
 
 #### Static type inference
-
+```typescript
+import v, { TypeOf } from 'validatex'
+
+const loginSchema = v.object({
+  username: v.string(),
+  password: v.string()
+})
+
+type LoginSchema = TypeOf<typeof loginSchema>
+// type LoginSchema = {
+//     username: string;
+//     password: string;
+// }
+```
 ___
 
 ## Built-In decoders
 #### String
+The **string** decoder parses unknown value, validates if its a string and returns an error or a string value on the basis of the validation.
 
+The signature of string decoder looks like:
+```typescript
+(options?: DecoderOption<string>) => (val: unknown, context?: Context | undefined) => string
+```
+
+```typescript
+const name = v.string()
+```
 
 ##### Validate
 
+```typescript
+function alphaNumeric(val: string) {
+  return /[a-zA-Z][\da-zA-Z]+/.test(val) ? undefined : 'Must be alpha numeric.';
+}
+
+const username = v.string(alphaNumeric);
+expect(username('user1')).toEqual('user1');
+// value validation
+expect(() => username('123')).toThrow('Must be alpha numeric.');
+```
+
 ##### Parse
+```typescript
+function parse(val: string) {
+  if (typeof val === 'string' || typeof val === 'number') return `${val}`;
+  return val;
+}
+
+const username = v.string(parse);
+
+expect(username('user1')).toEqual('user1');
+// parses value to string if its a number
+expect(username(123)).toEqual('123');
+```
 
 ##### Custom error message
+```typescript
+const customErrMsg = "This is a custom error message";
+
+const username = v.string(customErrMsg);
+
+expect(username.bind(username, 1)).toThrow(customErrorMsg);
+```
 
 #### Number
+The **number** decoder takes an unknown value, validates if its a number and returns an error or a number depending on the outcome of the validation.
+```typescript
+(options?: DecoderOption<number>) => (val: unknown, context?: Context | undefined) => number
```

Comment:
We need to document `DecoderOption` and `Context` too.

Reply:
Context is already documented on line 117, the code portion. Do I describe it? If i do then what should the description say.